### ISSUE TITLE
Check XML upload feature

### ISSUE THREAD

**Issue by dommitchell** [DOAJ/doaj#686]
![doaj eror](https://cloud.githubusercontent.com/assets/5955396/7293369/c31a2b82-e9a4-11e4-818e-8f2f342c49f9.jpg)
We have received emails this morning about the inability to upload XML. The majority of users have received this message:
'Failed to parse file - it is invalid XML; please fix it before attempting to upload again.'

and one received this:

S4s-elt-character: Non-whitespace Characters Are Not Allowed In Schema Elements Other Than 'xs:appinfo' And 'xs:documentation'. Saw '301 Moved Permanently'.

Can you check  that XML upload is working as it should? I am afraid that I forgot test this after our call on Monday.

**Comment by richard-jones** [comment]
Could you send us one or more of the broken XML uploads, and the user accounts?

The second error looks like a problem locating the schema - @emanuil-tolev could you check that the schema file is being served correctly by the new infrastructure?

**Comment by emanuil-tolev** [comment]
> The second error looks like a problem locating the schema - @emanuil-tolev could you check that the schema file is being served correctly by the new infrastructure?

Yes, of course. https://doaj.org/static/doaj/doajArticles.xsd

The 301 redirect is probably related to the fact we refer to the schema without HTTPS? But why wouldn't the XML validation follow the redirect, that's what a 301 Moved Permanently is for!

**Comment by richard-jones** [comment]
Could be this:

https://github.com/DOAJ/doaj/blob/production/portality/settings.py#L314

Check that path to schema on disk is correct

**Comment by richard-jones** [comment]
The second error must be caused by something different, but I'm not sure what.  I'd like to see the file that caused it.  

It may be that XSD validation won't follow redirects, for whatever reason.  I haven't been able to find anything online about the rules for that, but it could be one of them for all we know.

**Comment by emanuil-tolev** [comment]
> Check that path to schema on disk is correct

```python
>>> os.path.join(app.config['BASE_FILE_PATH'], "static", "doaj", "doajArticles.xsd")
'/home/cloo/repl/apps/doaj/src/doaj/portality/static/doaj/doajArticles.xsd'
```

```shell
$ ls -l /home/cloo/repl/apps/doaj/src/doaj/portality/static/doaj/doajArticles.xsd
-rw-rw-r-- 1 cloo cloo 4979 Apr 15 17:41 /home/cloo/repl/apps/doaj/src/doaj/portality/static/doaj/doajArticles.xsd
```

Yep, perfectly fine.

**Comment by emanuil-tolev** [comment]
Possibly relevant logs from XML ingestion in chronological order:

```
start timestamp 2015-04-22_173001
Running in production
Loaded final config from /home/cloo/repl/apps/doaj/src/doaj/production.cfg
processing upload record 9225fac280b84ed18502b753a2a52299
...downloaded as 9225fac280b84ed18502b753a2a52299.xml
validating 9225fac280b84ed18502b753a2a52299.xml
...failed
```

```
start timestamp 2015-04-23_060001
Running in production
Loaded final config from /home/cloo/repl/apps/doaj/src/doaj/production.cfg
processing upload record 673b4eaf10244afc94af6b6c4412fbca
...downloaded as 673b4eaf10244afc94af6b6c4412fbca.xml
validating 673b4eaf10244afc94af6b6c4412fbca.xml
...failed
processing upload record d5486c244ed74ebfba7e03c290e3a3fd
...downloaded as d5486c244ed74ebfba7e03c290e3a3fd.xml
validating d5486c244ed74ebfba7e03c290e3a3fd.xml
...failed
processing upload record 55f933438872421ab2273f97aafb5c9b
...downloaded as 55f933438872421ab2273f97aafb5c9b.xml
validating 55f933438872421ab2273f97aafb5c9b.xml
...failed
processing upload record 8f781faea10a45d9a3ac2c40f34e6cf9
...downloaded as 8f781faea10a45d9a3ac2c40f34e6cf9.xml
validating 8f781faea10a45d9a3ac2c40f34e6cf9.xml
...failed
```

```
start timestamp 2015-04-23_070001
Running in production
Loaded final config from /home/cloo/repl/apps/doaj/src/doaj/production.cfg
processing upload record c542c063359947d38e80669855866b47
...downloaded as c542c063359947d38e80669855866b47.xml
validating c542c063359947d38e80669855866b47.xml
...failed
processing upload record 4f398bf7da4246768da90d67a6421059
...downloaded as 4f398bf7da4246768da90d67a6421059.xml
validating 4f398bf7da4246768da90d67a6421059.xml
...failed
```

ALL of these are attempts to upload XML files via URL. Also the error most of these files got is pure "this is not even XML", before we get to any (other) validation.

This error:

> S4s-elt-character: Non-whitespace Characters Are Not Allowed In Schema Elements Other Than 'xs:appinfo' And 'xs:documentation'. Saw '301 Moved Permanently'.

is quite strange, since I'm not sure why *our* XML parser would be connecting to the internet to fetch the schema. We have it locally, it should use that, and it does use that.

**Comment by richard-jones** [comment]
ok, I've checked all these ids, and the first one is a link to a PDF, so that would explain why that isn't working.

The others look, superficially, to be correct so I'll need to experiment with them further.

Possibilities are:

1. They really aren't valid xml
2. Something goes wrong in the download process which corrupts the file
3. something else ... ?

@emanuil-tolev - are there any validation errors in the logs for uploaded files (and are there any uploaded files at all).  If so, could you let me know.

Also, are any of these files from your list above still on disk?  If so, could you send one over to me.

**Comment by richard-jones** [comment]
Taking this out of the PVOD milestone, so we can clear that up in parallel.  We haven't touched this code as part of that project, so cause is unknown, but could be related to move to new infrastructure.

**Comment by dommitchell** [comment]
> Could you send us one or more of the broken XML uploads, and the user accounts?

Sent by email

**Comment by emanuil-tolev** [comment]
> @emanuil-tolev - are there any validation errors in the logs for uploaded files (and are there any uploaded files at all).

Yep, the full logs are above in my comment https://github.com/DOAJ/doaj/issues/686#issuecomment-95649229 . The error ("... failed") is most likely caused by the exception raised at portality.article:626 , caught by portality.scripts.ingestarticles:172 . The handler in ingestarticles:172 is more general though, so it could theoretically be some other kind of exception in article:626, but I doubt it. The files just can't be parsed by lxml, they're invalid XML. (according to the info available so far)

> Also, are any of these files from your list above still on disk? If so, could you send one over to me.

No.

**Comment by emanuil-tolev** [comment]
> The files just can't be parsed by lxml, they're invalid XML. (according to the info available so far)

Yeah, hard data is proving this to be utter bollocks... Something is definitely wrong with upload, not the files, unless ALL the files so far have been invalid. For almost 3 days.

Last successful XML upload was at around 7pm on Tuesday before the move you see :).

https://doajgate.cottagelabs.com/doaj/upload/_search?sort=created_date:desc&q=status:processed

And there's more failures afterwards now.

I'm going to enable plain HTTP access to the schema again @richard-jones and see if uploads come through successfully on its old URL. So you will be able to GET http://doaj.org/static/doaj/doajArticles.xsd and you will not be redirected, simply given the file. HTTPS will stay enforced everywhere else. If that doesn't fix it then I'll revert the change and we should diagnose more.

**Comment by emanuil-tolev** [comment]
I've added more logging to the XML ingester so we get bit more details on the exception. I'll be at a bit of a loss if it's just a portality.article.IngestException, will really need serious digging then (but least we'll know the exact line of code causing the problem, and it's a line that uses an external library).

**Comment by dommitchell** [comment]
@richard-jones @emanuil-tolev Can I have an update on this please? Getting urgent now

**Comment by emanuil-tolev** [comment]
On it @dommitchell , we've just continued over email threads since it was going to get a bit much for github. I'm looking to resolve this entirely today. It's a very peculiar problem, not something I would have expected XML libraries to have trouble with.

**Comment by emanuil-tolev** [comment]
These all work without any redirects now, incl. over plain HTTP:

wget http://doaj.org/static/doaj/doajArticles.xsd
wget http://doaj.org/static/doaj/doajArticle.xsd
wget http://doaj.org/static/doaj/iso_639-2b.xsd

wget https://doaj.org/static/doaj/doajArticles.xsd
wget https://doaj.org/static/doaj/doajArticle.xsd
wget https://doaj.org/static/doaj/iso_639-2b.xsd

**Comment by emanuil-tolev** [comment]
(oh, FYI, doajArticle is actually a misspelling of doajArticles, but one that is so easy to make we have done it multiple times ourselves, so I've simply made it a synonym of doajArticle**s**)

**Comment by richard-jones** [comment]
This is now working - redirects have been made, local tests succeeding, and we can see a record on the live server of a successful upload.

**Comment by dommitchell** [comment]
Yep, user tested too. Thank you!

**Comment by dommitchell** [comment]
This seems to still be an issue. Now, users are getting this error:

"File could not be validated against a known schema; please fix this before attempting to upload again"

Has the schema location changed? Sample XML file sent via email.

**Comment by emanuil-tolev** [comment]
The last failed upload is at ~7:30am GMT today 4th May. The last successful upload finished processing at 8:03am GMT today 4th May.

This indicates there's no endemic problem anymore (before when this issue was opened there were no successes for several days straight). So this seems to be a separate issue - we'll see what it is.

> Has the schema location changed?

No, it's accessible in even more locations now in fact.

There are 2 schemas involved, doajArticles.xsd and iso_639-2b.xsd (for the language codes). They are both available directly without redirects over both secure and plain HTTP (see https://github.com/DOAJ/doaj/issues/686#issuecomment-97367707 for list of valid URLs). In addition a common misspelling of the doajArticles schema is available under the name doajArticle (singular).

These are all still working as they should (checked now).

There is a bit in the XML that is a little strange, ```xsi:noNamespaceSchemaLocation="https://doaj.org/static/doaj/doajArticles.xsd"```, leaving here for further investigation.

**Comment by dommitchell** [comment]
They don't need to reference the schema at all.
